aospy/automate.py

```diff
@@ -0,0 +1,220 @@
+                       'input_time_datatypes',
+                       'input_time_offsets',
+                       'input_vertical_datatypes',
+                       'output_time_intervals',
+                       'output_time_regional_reductions',
+                       'output_vertical_reductions')
+
+    _NAMES_SUITE_TO_CALC_MAPPING = {
+        'projects': 'proj',
+        'models': 'model',
+        'runs': 'run',
+        'variables': 'var',
+        'date_ranges': 'date_range',
+        'regions': 'region',
+        'input_time_intervals': 'intvl_in',
+        'input_time_datatypes': 'dtype_in_time',
+        'input_time_offsets': 'time_offset',
+        'input_vertical_datatypes': 'dtype_in_vert',
+        'output_time_intervals': 'intvl_out',
+        'output_time_regional_reductions': 'dtype_out_time',
+        'output_vertical_reductions': 'dtype_out_vert',
+    }
+
+    def __init__(self, calc_suite_specs, obj_lib):
+        """Turn all inputs into aospy-ready objects."""
+        self._specs_in = calc_suite_specs
+        self._obj_lib = obj_lib
+
+        self._projects = [self._obj_lib.projects[project_name]
+                          for project_name in self._specs_in['projects']]
+
+    def _permute_core_specs(self):
+        obj_trees = []
+        for project in self._projects:
+            models = _get_mult_objs_by_names(self._specs_in['models'], project,
+                                             'models', 'default_models')
+            for model in models:
+                runs = _get_mult_objs_by_names(self._specs_in['runs'], model,
+                                               'runs', 'default_runs')
+                for run in runs:
+                    # Use Calc's attr names as keys.
+                    obj_trees.append(dict(proj=project, model=model, run=run))
+        return obj_trees
+
+    def _permute_aux_specs(self):
+        # Drop the "core" specifications, which are handled separately.
+        specs = self._specs_in.copy()
+        [specs.pop(core) for core in self._CORE_SPEC_NAMES]
+        # TODO: don't just pull from first project.
+        regions = _get_mult_objs_by_names(specs['regions'], self._projects[0],
+                                           'regions', 'regions')
+        variables = _get_mult_objs_by_names(specs['variables'], self._obj_lib,
+                                            'variables', 'variables')
+        specs['regions'] = regions
+        specs['variables'] = variables
+
+        # Convert to attr names that Calc is expecting.
+        calc_aux_mapping = self._NAMES_SUITE_TO_CALC_MAPPING.copy()
+        [calc_aux_mapping.pop(core) for core in self._CORE_SPEC_NAMES]
+        for suite_name, calc_name in calc_aux_mapping.items():
+            specs[calc_name] = specs.pop(suite_name)
+        return _permute_aux_specs(specs)
+
+    def _combine_core_aux_specs(self):
+        return _combine_core_aux_specs(self._permute_core_specs(),
```

Comment:
I agree, I got carried away here with the module-level functions.  The original motivation was for ease of testing and potential re-use of functionality elsewhere, but this logic seems pretty specific to CalcSuite.